source/documentation/adrs/adr-009.html.md.erb

```diff
@@ -0,0 +1,68 @@
+---
+title: ADR-009 Log structure
+last_reviewed_on: 2024-02-09
+review_in: 12 months
+---
+# <%= current_page.data.title %>
+
+Date: 14-02-2024
+
+## Status
+
+🤔 Proposed
+
+## Context
+
+Regardless of platform or language, our services write logging information to the output stream. However, this logging information is inconsitently structured. This leads to a few problems:
+
+1. It is hard to interpret logs under standard operation, as logs from different containers with different structures are interspersed
+2. During an incident, it is hard to find related logs as the inconsistent structures require separate queries
+3. Our log-based alarms are complex to write and update as they have to query several different structures
+
+## Decision
+
+We will define a standard logging structure that all teams will use. Teams are welcome to expand on the structure by adding extra fields, but must include all required fields in each log message and must ensure not to use reserved fields for anything other than their defined usage.
+
+All logs must be formatted as JSON objects.
+
+An example of a minimal log message:
+
+```json
+{"time": "2024-02-14T12:34:23Z", "level": "CRITICAL", "msg": "Null pointer exception"}
+```
+
+An example of a more complex log message:
+
+```json
+{
+  "time": "2024-02-14T13:39:01Z",
+  "level": "INFO",
+  "msg": "User permissions updated",
+  "trace_id": "1-581cf771-a006649127e371903a2de979",
+  "request": { "method": "PUT", "path": "/user/133/permissions" },
+  "location": { "file": "pages/user/edit_permission.go", "line": 156 },
+  "actor_id": 48
+}
+```
+
+As well as the required and reserved fields below, log messages can include any other fields (such as "location" and "actor_id" above).
+
+### Required fields
```

Comment:
No required field for service? I feel like I always want to know which service emitted a log to avoid confusion and ambiguity.

Reply:
Charles mentioned this [on Slack too](https://mojdt.slack.com/archives/CCM1S4B8F/p1708008005143069?thread_ts=1708002867.593279&cid=CCM1S4B8F), and I originally had a `service` field, but it became hard to define because I'm not clear of a consistent definition of what a service is, or what level of detail would be helpful (e.g. "sirius" vs "uid-service" vs "uid-service-create").

To avoid getting into a longer discussion about it, I suggest it can be left to teams to determine what level of naming/differentiation is best for them.

If we do have a clear naming convention though I'll happily add it back in.

Reply:
Maybe there's a gap for "recommended fields" then. I'd generally use service in this context to mean the meaningful name of the container or lambda outputting the log in context of the app in question.

Reply:
I think I've let perfect be the enemy of good here and there's clearly value in having _something_ to represent source. I've stolen the core of your definition there and added it in as `service_name`, which is the name we already use in our Go apps.